onedal/backend/data/data.cpp

```diff
@@ -0,0 +1,194 @@
+
+#include <Python.h>
+#include "backend/data/data.h"
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/row_accessor.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#ifdef ONEDAL_DATA_PARALLEL
+#include <CL/sycl.hpp>
+#endif
+
+#ifdef DPCTL_ENABLE
+#include "dpctl_sycl_types.h"
+#include "dpctl_sycl_queue_manager.h"
+#endif
+
+namespace oneapi::dal::python {
+#define is_array(a)         ((a) && PyArray_Check(a))
+#define array_type(a)       PyArray_TYPE((PyArrayObject *)a)
+#define array_is_behaved(a) (PyArray_ISCARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_behaved_F(a) \
+    (PyArray_ISFARRAY_RO((PyArrayObject *)a) && array_type(a) < NPY_OBJECT)
+#define array_is_native(a) (PyArray_ISNOTSWAPPED((PyArrayObject *)a))
+#define array_numdims(a)   PyArray_NDIM((PyArrayObject *)a)
+#define array_data(a)      PyArray_DATA((PyArrayObject *)a)
+#define array_size(a, i)   PyArray_DIM((PyArrayObject *)a, i)
+
+#ifdef ONEDAL_DATA_PARALLEL
+int init_numpy() {
+    import_array();
+    return 0;
+}
+
+const static int numpy_initialized = init_numpy();
+#endif
+
+class numpy_deleter {
+public:
+    numpy_deleter(PyArrayObject *a) : ndarray_(a) {}
+
+    void operator()(const void *ptr) {}
+
+    numpy_deleter &operator=(const numpy_deleter &) = delete;
+
+private:
+    PyArrayObject *ndarray_;
+};
+
+template <typename T, typename ConstDeleter>
+inline dal::homogen_table create_homogen_table(const T *data_pointer,
+                                               const std::size_t row_count,
+                                               const std::size_t column_count,
+                                               const dal::data_layout layout,
+                                               ConstDeleter &&data_deleter) {
+    return dal::homogen_table(data_pointer, row_count, column_count, data_deleter, layout);
+}
+
+template <typename T>
+inline dal::homogen_table make_homogen(PyArrayObject *array) {
+    size_t column_count = 1;
+
+    if (array_numdims(array) > 2) {
+        throw std::runtime_error("Input array has wrong dimensionality (must be 2d).");
+    }
+    T *data_pointer = reinterpret_cast<T *>(array_data(array));
+    const size_t row_count = static_cast<size_t>(array_size(array, 0));
+    if (array_numdims(array) == 2) {
+        column_count = static_cast<size_t>(array_size(array, 1));
+    }
+    const auto layout =
+        array_is_behaved_F(array) ? dal::data_layout::column_major : dal::data_layout::row_major;
+    auto res_table =
+        create_homogen_table(data_pointer, row_count, column_count, layout, numpy_deleter(array));
+    // we need it increment the ref-count if we use the input array in-place
+    // if we copied/converted it we already own our own reference
+    if (reinterpret_cast<PyArrayObject *>(data_pointer) == array)
+        Py_INCREF(array);
+    return res_table;
+}
+
+dal::table input_to_onedal_table(PyObject *obj) {

```

Comment:
`convert_to_table()`

Reply:
renamed to `convert_from_numpy_to_table` and `convert_from_table_to_numpy`.  It's ok?